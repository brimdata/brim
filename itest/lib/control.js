"use strict";
exports.__esModule = true;
exports.retryUntil = exports.retry = void 0;
function retry(f, attempts, delay) {
    if (attempts === void 0) { attempts = 100; }
    if (delay === void 0) { delay = 100; }
    return new Promise(function (resolve, reject) {
        f().then(function (ret) { return resolve(ret); })["catch"](function (err) {
            setTimeout(function () {
                if (attempts === 1) {
                    reject(err);
                }
                else {
                    retry(f, attempts - 1, delay).then(function (ret) { return resolve(ret); })["catch"](function (err) { return reject(err); });
                }
            }, delay);
        });
    });
}
exports.retry = retry;
exports.retryUntil = function (f, cond_f, attempts, delay) {
    if (attempts === void 0) { attempts = 15; }
    if (delay === void 0) { delay = 1000; }
    // it is resolved and is responsible for examining and returning a boolean to
    // determine if the condition is satisfied.
    return retry(function () { return new Promise(function (resolve, reject) {
        f().then(function (val) {
            if (cond_f(val)) {
                resolve(val);
            }
            else {
                reject(new Error("retryUntil condition failure: " + val));
            }
        })["catch"](function (err) {
            reject("retryUntil promise failure: " + err);
        });
    }); }, attempts, delay);
};
