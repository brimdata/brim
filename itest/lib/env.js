/* @flow */

import path from "path"

export const isCI = (): boolean => process.env.GITHUB_ACTIONS === "true"

export const repoDir = (): string =>
  path.resolve(path.join(__dirname, "..", ".."))

export const itestDir = (): string =>
  path.join(process.env.WORKSPACE || "run", "itest")

export const testDataDir = (): string =>
  path.resolve(path.join(repoDir(), "itest", "testdata"))
