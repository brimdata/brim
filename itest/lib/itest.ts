import startApp from "./appStep/api/startApp"
import BrimDriver from "./brim-driver"
import BrimSelectors from "./brim-selectors"
import newAppInstance from "./newAppInstance"

export default function itest(name): [BrimDriver, BrimSelectors] {
  let app
  let testIdx = 0
  let brim = new BrimDriver(app)
  let $ = new BrimSelectors()

  beforeAll(async () => {
    app = await newAppInstance(name, ++testIdx)
    brim.app = app
    await startApp(app)
  })

  afterAll(async () => {
    if (app && app.isRunning()) await app.stop()
  })

  return [brim, $]
}
