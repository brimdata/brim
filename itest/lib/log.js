"use strict"
exports.__esModule = true
exports.LOG = void 0
var path_1 = require("path")
var winston_1 = require("winston")
var env_1 = require("./env")
// When the flow-typed update produces an error....
// flow-typed/npm/winston_v3.x.x.js
// [1] 124â”‚     createLogger: <T>($winstonLoggerConfig<T>) => $winstonLogger<T>,
exports.LOG = winston_1.createLogger({
  level: "info",
  format: winston_1.format.combine(
    winston_1.format.timestamp({
      format: "YYYY-MM-DD HH:mm:ss.SSS"
    }),
    winston_1.format.printf(function(_a) {
      var level = _a.level,
        message = _a.message,
        timestamp = _a.timestamp
      return "[" + timestamp + " " + level + "]: " + message
    })
  ),
  transports: [
    new winston_1.transports.Console({level: "info"}),
    // mkdir -p semantics so the environment variable seems OK enough.
    new winston_1.transports.File({
      filename: path_1["default"].join(env_1.itestDir(), "itest.log"),
      level: "debug"
    })
  ]
})
