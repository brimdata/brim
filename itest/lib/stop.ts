import fetch from "node-fetch"

export async function quitBrim(app) {
  if (app && app.isRunning()) {
    await app.stop()
    console.log("app stopped, waiting for zqd....")
    while (await isUp()) {
      await new Promise((go) => setTimeout(go, 500))
    }
    console.log("zqd is down")
  }
}

async function isUp() {
  try {
    const response = await fetch("http://localhost:9867/status")
    const text = await response.text()
    return text === "ok"
  } catch (e) {
    return false
  }
}
