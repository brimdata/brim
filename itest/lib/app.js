/* @flow */

import {retry} from "./control"
import {selectors} from "../../src/js/test/integration"
import {Application} from "spectron"

export const logIn = (app: Application) => {
  return app.client
    .setValue(selectors.login.host, "localhost")
    .setValue(selectors.login.port, "9867")
    .click(selectors.login.button)
}

export const waitForLoginAvailable = (app: Application) => {
  const waitForHostname = () => {
    return app.client.waitForExist(selectors.login.host)
  }
  const waitForPort = () => {
    return app.client.waitForExist(selectors.login.port)
  }
  const waitForButton = () => {
    return app.client.waitForExist(selectors.login.button)
  }
  return waitForHostname()
    .then(() => waitForPort())
    .then(() => waitForButton())
}

export const waitForSearch = (app: Application) => {
  return retry(() => app.client.element("#main-search-input").getValue())
}

export const waitForHistogram = (app: Application) => {
  return retry(() =>
    app.client.element(selectors.histogram.topLevel).getAttribute("class")
  )
}

export const writeSearch = (app: Application, searchText: string) =>
  app.client.setValue(selectors.search.input, searchText)

export const startSearch = (app: Application) =>
  app.client.click(selectors.search.button)

export const searchDisplay = async (app: Application) => {
  // This stinks. We have to use getHTML because headers that are off the
  // screen return as empty strings if you use getText. This isn't required of
  // actual results.
  // See http://v4.webdriver.io/api/property/getText.html
  // app.browserWindow.maximize() fixes the problem on my laptop but not CircleCI.
  // But what we get back includes the width which can be non-deterministic:
  // <div class="header-cell" style="width: 192px;">ts<div class="col-resizer"></div></div>
  // That style width will vary on my laptop vs. CircleCI.
  // The hack is to split this and extract just the text.
  const _trim = (s: string) => s.split(">")[1].split("<")[0]

  const headerResults = () => {
    return app.client.getHTML(selectors.viewer.headers).then((headers) => {
      if (typeof headers === "string") {
        headers = [headers]
      }
      return headers.map((h) => _trim(h))
    })
  }
  const searchResults = () => app.client.getText(selectors.viewer.results)

  let headers = await headerResults()
  let search = await searchResults()
  return headers.concat(search)
}

export const getCurrentSpace = (app: Application) =>
  app.client.getText(selectors.spaces.button)
