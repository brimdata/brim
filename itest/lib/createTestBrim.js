/* @flow */
import {
  click,
  ingestFile,
  newAppInstance,
  startApp,
  startSearch,
  writeSearch
} from "./app"
import {retryUntil} from "./control"
import lib from "../../src/js/lib"

export default (name: string) => {
  let app
  let testIdx = 0

  beforeAll(async () => {
    app = await newAppInstance(name, ++testIdx)
    await startApp(app)
  })

  afterAll(async () => {
    if (app && app.isRunning()) await app.stop()
  })

  return {
    ingest(file: string) {
      return ingestFile(app, file)
    },

    mockSaveDialog(value: {canceled: boolean, filePath: ?string}) {
      return app.mainProcess.emit(
        "spectron:mock",
        "windows:showSaveDialog",
        value
      )
    },

    writeSearch(input: string) {
      return writeSearch(app, input)
    },

    startSearch() {
      return startSearch(app)
    },

    click(locator: string) {
      return click(app, locator)
    },

    waitForText(locator: string, regex: RegExp) {
      return retryUntil(
        () => app.client.getText(locator),
        (s) => regex.test(s)
      )
    },

    wait(ms: number) {
      return lib.sleep(ms)
    }
  }
}
