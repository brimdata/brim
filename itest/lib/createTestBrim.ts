import {Locator} from "../../src/js/test/createLocator"
import appStep from "./appStep/api"
// TODO in a future PR: remove direct logStep uses here.
import logStep from "./appStep/util/logStep"
import newAppInstance from "./newAppInstance"
import {retryUntil} from "./control"
import lib from "../../src/js/lib"

export default (name: string) => {
  let app
  let testIdx = 0

  beforeAll(async () => {
    app = await newAppInstance(name, ++testIdx)
    await appStep.startApp(app)
  })

  afterAll(async () => {
    if (app && app.isRunning()) await app.stop()
  })

  return {
    getApp() {
      return app
    },

    ingest(file: string) {
      return appStep.ingestFile(app, file)
    },

    mockSaveDialog(value: {
      canceled: boolean
      filePath: string | null | undefined
    }) {
      return app.mainProcess.emit(
        "spectron:mock",
        "windows:showSaveDialog",
        value
      )
    },

    setValue(locator: Locator, value: string) {
      return logStep(`set input ${locator.css} to ${value}`, async () => {
        const el = await app.client.$(locator.css)
        await el.waitForDisplayed()
        return el.setValue(value)
      })
    },

    getText(locator: Locator) {
      return logStep(`get text from ${locator.css}`, async () => {
        const el = await app.client.$(locator.css)
        await el.waitForDisplayed()
        return el.getText()
      })
    },

    clickAppMenuItem(id: string) {
      return app.mainProcess.emit("spectron:clickAppMenuItem", id)
    },

    search(input: string) {
      return appStep.search(app, input)
    },

    click(locator: Locator) {
      return appStep.click(app, locator.css)
    },

    waitForText(locator: string, regex: RegExp) {
      return retryUntil(
        async () => (await app.client.$(locator)).getText(),
        (s) => regex.test(s)
      )
    },

    async isVisible(locator: Locator) {
      return !!(await (await app.client.$(locator.css)).isDisplayed())
    },

    async isNotVisible(locator: Locator) {
      return !(await (await app.client.$(locator.css)).isDisplayed())
    },

    wait(ms: number) {
      return lib.sleep(ms)
    }
  }
}
