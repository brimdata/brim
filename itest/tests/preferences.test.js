/* @flow */
import path from "path"

import {
  defaultModalButton,
  ingestProgressBar,
  jsonTypeConfigInput,
  viewerResults
} from "../../src/js/test/locators"
import {stdTest} from "../lib/jest"
import createTestBrim from "../lib/createTestBrim"
import testDataDir from "../lib/env"

const config = path.join(testDataDir(), "custom-schema.json")

describe("Preferences Modal Tests", () => {
  const brim = createTestBrim("preferences.test")

  stdTest("setting a custom json type config", async (done) => {
    brim.clickAppMenuItem("preferences")
    await brim.setValue(jsonTypeConfigInput, config)
    await brim.click(defaultModalButton)
    await brim.ingest("custom-sample.ndjson")
    await brim.waitUntil(() => brim.isNotVisible(ingestProgressBar))

    expect(await brim.getText(viewerResults)).toMatchSnapshot()
    done()
  })
})
