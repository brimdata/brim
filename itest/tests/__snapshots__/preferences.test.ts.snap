// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preferences Modal Tests setting a custom json type config 1`] = `
Array [
  "ts",
  "src_ip",
  "src_port",
  "dst_ip",
  "dst_port",
  "proto",
  "2020-02-25T16:03:05.983",
  "192.168.1.110",
  "56625",
  "172.217.9.142",
  "443",
  "tcp",
  "2020-02-25T16:03:06.154",
  "192.168.1.110",
  "57591",
  "172.217.9.142",
  "443",
  "tcp",
  "2020-02-25T16:03:07.715",
  "192.168.1.110",
  "55635",
  "18.246.31.137",
  "443",
  "tcp",
  "2020-02-25T16:03:07.715",
  "192.168.1.110",
  "55747",
  "13.52.5.22",
  "443",
  "tcp",
  "2020-02-25T16:03:08.210",
  "192.168.1.110",
  "55344",
  "52.37.243.173",
  "443",
  "tcp",
]
`;
